src/pgtable.c

```diff
@@ -0,0 +1,602 @@
+
+#define pgt_entry_test(_pte, _flag) \
+	((_pte)->value & (_flag))
+
+#define pgt_entry_present(_pte) \
+	pgt_entry_test(_pte, PGT_ENTRY_FLAG_REGION | PGT_ENTRY_FLAG_DIR)
+
+#define pgt_is_addr_aligned(_addr) \
+	(!((_addr) & (PGT_ADDR_ALIGN - 1)))
+
+#define pgt_check_ptr(_ptr) \
+	do { \
+		assert(!((uintptr_t)(_ptr) & (PGT_ENTRY_MIN_ALIGN - 1))); \
+	} while (0)
+
+#define pgt_entry_set_region(_pte, _region) \
+	do { \
+		pgt_region_t *tmp = (_region); \
+		pgt_check_ptr(tmp); \
+		(_pte)->value = ((uintptr_t)tmp) | PGT_ENTRY_FLAG_REGION; \
+	} while (0)
+
+#define pgt_entry_set_dir(_pte, _dir) \
+	do { \
+		pgt_dir_t *tmp = (_dir); \
+		pgt_check_ptr(tmp); \
+		(_pte)->value = ((uintptr_t)tmp) | PGT_ENTRY_FLAG_DIR; \
+	} while (0)
+
+#define pgt_entry_get_region(_pte) \
+	({ \
+		assert(pgt_entry_test(_pte, PGT_ENTRY_FLAG_REGION)); \
+		(pgt_region_t*)pgt_entry_value(_pte); \
+	})
+
+#define pgt_entry_get_dir(_pte) \
+	({ \
+		assert(pgt_entry_test(_pte, PGT_ENTRY_FLAG_DIR)); \
+		(pgt_dir_t*)pgt_entry_value(_pte); \
+	})
+
+
+static inline pgt_dir_t* pgt_dir_alloc(pgtable_t *pgtable)
+{
+	pgt_dir_t *pgd;
+
+	pgd = pgtable->pgd_alloc_cb(pgtable);
+	if (pgd == NULL) {
+		fprintf(stderr, "Failed to allocate page table directory");
+	}
+
+	pgt_check_ptr(pgd);
+	memset(pgd, 0, sizeof(*pgd));
+	return pgd;
+}
+
+static inline void pgt_dir_release(pgtable_t *pgtable, pgt_dir_t* pgd)
+{
+	pgtable->pgd_release_cb(pgtable, pgd);
+}
+
+static inline void pgt_address_advance(pgt_addr_t *address_p,
+										   unsigned order)
+{
+	assert(order < 64);
+	/* coverity[large_shift] */
+	*address_p += 1ul << order;
+}
+
+static void pgtable_reset(pgtable_t *pgtable)
+{
+	pgtable->base  = 0;
+	pgtable->mask  = ((pgt_addr_t)-1) << PGT_ADDR_SHIFT;
+	pgtable->shift = PGT_ADDR_SHIFT;
+}
+
+/**
+ * Make the page table map a wider range of addresses - expands by PGT_ENTRY_SHIFT.
+ */
+static void pgtable_expand(pgtable_t *pgtable)
+{
+	pgt_dir_t *pgd;
+
+	assert(pgtable->shift <= (PGT_ADDR_ORDER - PGT_ENTRY_SHIFT));
+
+	if (pgt_entry_present(&pgtable->root)) {
+		pgd = pgt_dir_alloc(pgtable);
+		pgd->entries[(pgtable->base >> pgtable->shift) & PGT_ENTRY_MASK] =
+						pgtable->root;
+		pgd->count = 1;
+		pgt_entry_set_dir(&pgtable->root, pgd);
+	}
+
+	pgtable->shift += PGT_ENTRY_SHIFT;
+	pgtable->mask <<= PGT_ENTRY_SHIFT;
+	pgtable->base  &= pgtable->mask;
+}
+
+/**
+ * Shrink the page table address span if possible
+ *
+ * @return Whether it was shrinked.
+ */
+static int pgtable_shrink(pgtable_t *pgtable)
+{
+	pgt_entry_t *pte;
+	pgt_dir_t *pgd;
+	unsigned pte_idx;
+
+	if (!pgt_entry_present(&pgtable->root)) {
+		pgtable_reset(pgtable);
+		return 0;
+	} else if (!pgt_entry_test(&pgtable->root, PGT_ENTRY_FLAG_DIR)) {
+		return 0;
+	}
+
+	pgd = pgt_entry_get_dir(&pgtable->root);
+	assert(pgd->count > 0); /* should be empty */
+
+	/* If there is just one PTE, we can reduce the page table to map
+	 * this PTE only.
+	 */
+	if (pgd->count != 1) {
```

Comment:
No need for braces over a single statement, unless for matching `if ... else ...` structure. There are many similar occurrences in the file.

Reply:
@j-xiong not the author but I personally prefer braces over single line if statements anyway. Leads to less accidents when adding new lines to the if statements (adding a new line w/o braces will lead to some hard to find bugs but it will still compile)

Reply:
@wckzhang Although that is personal preference, the coding convention here is to not use braces over a single line. Braces are added when new lines are added -- not too difficult to follow.

No coding convention can eliminate coding accident. Sticking with the same convention does help improve code readability in the long run.